### PR TITLE
fix(telegram): keep fenced session transports attached

### LINKED ISSUE #4052: Telegram notifications fail despite healthy daemon after 0.12.16
## Summary

Telegram notifications still fail on `0.12.16` after the merged authority-outage fix in #3974.

## Observed

A user reproduced the same behavior on both `0.12.15` and `0.12.16`: the daemon appears healthy, but cannot connect to the session endpoint, so Telegram notifications are not delivered. `gjc daemon reload` and a stop/restart did not restore delivery. The owner independently confirmed the failure on `0.12.16`.

## Scope

Investigate the current release/runtime path rather than assuming #3974's shared-topic-authority crash fix covers this symptom. Establish a minimal reproducible endpoint-connectivity failure; preserve safe handling of tokens and private configuration.

## Acceptance criteria

- Reproduce or precisely classify the endpoint connection failure on current `dev` / `0.12.16`.
- Identify whether the failure is daemon lifecycle, endpoint discovery, status health reporting, or configuration/runtime-specific.
- Add a regression test if a repository defect is found.
- Keep the issue update public-safe; logs/configuration must be redacted.

### PR DESCRIPTION
**Pull request by chiznoir**
## Summary

- keep an authenticated same-session Telegram transport attached while its predecessor topic archive converges
- settle Telegram's definitive non-forum close response instead of retrying an impossible archive forever
- publish daemon generation 59 and refresh guarded manifests

Closes #4052.

## Evidence

- Reproduced a healthy daemon with one live endpoint repeatedly dropping the session as `recovery_rejected`, then idle-exiting with a stale heartbeat.
- Confirmed Telegram returned `400: Bad Request: the chat is not a supergroup forum` while the registry remained `archive_pending`.
- Redacted reproduction and verification receipt: `artifacts/telegram-daemon-reproduction-2026-08-09.json`.

## Tests

- `bun test packages/coding-agent/test/notifications-telegram-daemon.test.ts packages/coding-agent/test/notifications-topic-registry.test.ts` — 616 pass
- coding-agent Biome and TypeScript checks — pass
- Telegram generation guard — pass
- Telegram baseline manifest — 60 command receipts pass
- current `upstream/dev` local source daemon — 180-second sustained heartbeat, health OK, one live endpoint, Telegram delivery OK
- binary and native builds — pass

## Local full-suite notes

- `bun run check` reached the SDK downgrade fixture after Telegram gates passed; that fixture exceeded its 300-second native-build timeout.
- `bun run test` also exposed unrelated baseline/environment failures (missing local `cargo-nextest`, AI evidence drift, TUI expectations, macOS `/var` canonicalization, SDK fixture races, and MCP warning counts) before the coding-agent suite was SIGKILLed.

### CODE REVIEW

File: packages/coding-agent/src/sdk/bus/topic-registry.ts [hunk 1]
```diff
@@ -303,12 +303,19 @@ export function parseTopicRegistryState(value: unknown): TopicRegistryState | un
...
+			if (
+				raw.authorityState !== "archive_pending" &&
+				raw.authorityState !== "archive_exhausted" &&
+				raw.authorityState !== "inactive"
+			)
+				malformed();
+			// Older archive transitions retained the expired disconnect-grace marker,
+			// making the entire shared registry unreadable on the next daemon start.
+			delete raw.disconnectGraceExpiresAt;
+		}
```
[the remark this answers is not in the document]
Comment (chiznoir) [reply packages/coding-agent/src/sdk/bus/topic-registry.ts:338]: Fixed in 4780444d5. `parseTopicRegistryState` now clones the top-level state and each topic record before normalization, and the regression test asserts that the caller-provided snapshot remains unchanged. Focused Telegram suites and the coding-agent check pass.

**Comment by chiznoir** [comment]
Required CI is currently blocked by the same 12 shard-1 failures already present on the exact `dev@6608c2014` baseline run `31342922574`, before this PR’s Telegram-only diff. The ten continuation failures are tracked in #4124, and the two model-default/fork-context failures are tracked in #4125; #4125 also records PR #4117 as non-causal comparison evidence. Telegram-specific checks, generation guard, Windows safety, package checks, native builds, and focused notification suites pass. This PR will be rebased and rerun after the baseline repairs land.

**Review by Yeachan-Heo — DISMISSED** [review]
## REQUEST_CHANGES — adversarial exact-head review of `b5eea249cf726146dc8d72567e1c123c6d37c78d`

Reviewed at the exact head against PR base `b4c4f6044c` and current `dev@99e9758774`. Everything below was reproduced locally in a detached worktree at the exact head (Bun 1.3.14, linux-x64, natives built from `crates/pi-natives` at head). No mutation was made to this branch.

### What holds up

The core design is sound and I verified it end to end, not just by reading it.

- **Focused suites at head:** `notifications-telegram-daemon.test.ts` 558 pass / 0 fail, `notifications-topic-registry.test.ts` 58 pass / 0 fail (= the 616 you claim), `scripts/telegram-daemon-generation-guard.test.ts` 48 pass / 0 fail, `bun --cwd=packages/coding-agent run check` clean (biome 2654 files + `tsc --noEmit`).
- **Generation guard:** `bun scripts/telegram-daemon-generation-guard.ts b4c4f6044c b5eea249cf` → `v43 required generation bump verified`. `DAEMON_GENERATION` 59 → 60 stacks correctly on #4128 (whose merge commit *is* this PR's base), and the manifest deltas (`DAEMON_GENERATION`, `handleSessionMessage`) match the guard's protected inventory; `topicArchiveSettled`, `scanRoots`, and `topic-registry.ts` are outside that inventory, so the two-hash manifest diff is correct rather than under-updated.
- **Windows safety:** the win32 job can't run here, but the equivalent name-pattern subset passes locally (9 pass / 0 fail), and the job itself is green on run `31350094746`.
- **Convergence actually converges.** I wrote an adversarial probe that drives the full outage path: `archive_pending` fence → attached fenced transport with a `rejected` recovery lease → archive settles → `scanRoots` reconciles to `inactive`, redials, and the successor transport bootstraps a *new* topic with delivery restored. It passes at head. That is the real repair for #4052 and it works.
- **Fencing is not weakened where it matters.** Mutation testing killed 7 of 9 mutants, including the two I most wanted to break: forcing `gapAllowsRecovery = true` (accept any gap) is caught by two tests, and dropping `replayCandidateSessionId === session.sessionId` from the new archive branch is caught — so the cross-session leak and the over-permissive replay-gap admission are both genuinely fenced. Registry parsing is now side-effect free and the stale disconnect-grace repair is covered.

### Blocking

**1. The non-forum settlement only matches one of Telegram's two phrasings, and this file already knows the other one.**

`topicArchiveSettled` (`telegram-daemon.ts:534`) adds an *anchored* alternative `the chat is not a supergroup forum`. But `telegram-daemon.ts:435` — same file, capability detection — already treats `chat is not a forum` as the non-forum signal. Reproduced by swapping only the description in your own parameterized archive-reconciliation table:

```
description: "Bad Request: the chat is not a forum"
→ expected authorityState "inactive", got "archive_pending" (+ a persisted retry job)
```

That is the exact symptom this PR exists to kill: an impossible close retried forever against a chat that can never have topics. Broaden the alternation (e.g. `the chat is not a (?:supergroup )?forum`) and add the row to the table beside the one you just added.

**2. The convergence trigger is load-bearing and has zero coverage.**

Deleting `recoveryRejectedAfterArchive ||` from the `scanRoots` reconnect predicate leaves **all 616 focused tests green**. That one clause is the only thing that ever redials a session whose recovery was rejected while its predecessor archived — `#recoverTopicBinding` is called from exactly one place (the replay barrier), so without the redial an attached transport that took the new keep-attached branch is permanently mute even after the archive settles. My probe fails under that mutation (`successor` is still the fenced session) and under removal of the keep-attached branch itself, so a single regression test pins both halves. Please land that test with the fix; the recipe is: fenced archive → attach → settle → `scanRoots` → assert a *new* session object, one new `createForumTopic`, and restored delivery.

### Non-blocking, but fix while you're here

3. `topic-registry.ts:1226` — `disconnectGraceExpiresAt: undefined` inside the `restoreArchiveFence` object literal is a no-op: lines 1228–1229 immediately `delete` that key on the same object. Removing it survives every test. Drop one of the two.

4. The keep-attached branch clears `session.replayPending = false` and returns without doing either thing the success path does at the same point: it never drains `session.replayQueue`, and it skips the `#ownsLiveOpenEndpoint` / `#leaseAllows` revalidation at lines 10037–10043. Reproduced the first one — a `turn_stream` parked behind the barrier is still sitting in `replayQueue` after the barrier clears, and is discarded when `scanRoots` replaces the session. Delivery impact is nil today because the fence drops those frames anyway, but the barrier contract in the comment two branches up says the queue is never left half-open, and re-running the ownership check before clearing an admission barrier costs nothing.

5. Evidence hygiene: `artifacts/telegram-daemon-reproduction-2026-08-09.json` records `testedHead: e028997f4`, which is the *parent* of the PR head. I verified `git diff e028997f47 b5eea249cf -- packages crates` is empty (only the manifest and the artifact changed), so the dogfood evidence is source-equivalent and I'm treating it as valid — but the field should say `b5eea249c`. The artifact is otherwise clean: no tokens, chat ids, usernames, or message bodies, and its claims (generation 60, non-forum 400, same-session retention, 616 focused tests) match the diff and my reruns.

### CI attribution — please update the standing comment

Your comment claims 12 shard-1 failures from an older baseline. At the exact head, run `31350094746` has exactly **two**:

- `ModelSelector assignment menu role bindings > reports the effective default when an active profile supplies it`
- `fork context policy surface > passes a sanitized fork seed and cache identity without sharing provider state`

I reran both on `dev@99e9758774` with no PR content: **fork-context now passes** — #4134 repaired it — and only the ModelSelector effective-default failure remains, still failing without this PR. So after the rebase the honest claim is "one pre-existing non-causal dev failure", not twelve. Every Telegram-relevant job (both notification suites, the generation guard, Windows Telegram daemon safety) is green at this head.

### Landing order

- **#4128 is already satisfied:** its merge commit `b4c4f6044c` is this PR's base, so the 59 → 60 generation stack is correct and there is no supersession question.
- **#4135 is unrelated:** its 14 files are provider presets / onboarding / `schemas/config.schema.json`; it touches no `sdk/bus` or notification file. No ordering constraint in either direction.
- **Rebase onto `99e9758774` is conflict-free** (verified with `git merge-tree`), so rebase, rerun, and the only expected red is the ModelSelector regression owned by dev.

Land order: rebase onto current dev → fix (1) and (2) → this can go in independently of #4135.

Not merging; re-review on the next head.

—
*[repo owner's gaebal-gajae (clawdbot) 🦞]*

**Comment by Yeachan-Heo** [comment]
Head/base tracking update — `dev` moved again during this review (`99e9758774` → `4ad03a5fce`, #4140), so the CI attribution in my review is now stale in your favour.

- #4140 stubs `getModelProfile` in the model-selector registry fake, which is the last of the two shard-1 failures on your exact head. I reran both offenders on `dev@4ad03a5fce` with no PR content: **39 pass / 0 fail** — fork-context fixed by #4134, ModelSelector effective-default fixed by #4140.
- `git merge-tree` against `4ad03a5fce` is still conflict-free for head `b5eea249cf`, and #4140 touches only `packages/coding-agent/test/model-selector-action-menu-role-binding.test.ts`.

So after a rebase onto current dev there is no longer any pre-existing failure to attribute: shard-1 should be fully green, and the only remaining blockers are the two in my review — the phrase-narrow non-forum settlement (`the chat is not a forum` still never settles) and the missing regression test for the `recoveryRejectedAfterArchive` convergence trigger. PR head is unchanged at `b5eea249cf726146dc8d72567e1c123c6d37c78d`; re-review on the rebased head.

—
*[repo owner's gaebal-gajae (clawdbot) 🦞]*

**Comment by Yeachan-Heo** [comment]
Amendment to my REQUEST_CHANGES review — a second reviewer pass found one defect I missed and one factual error in my own text. Head is unchanged at `b5eea249cf726146dc8d72567e1c123c6d37c78d`; the REQUEST_CHANGES verdict stands and now has a stronger blocker.

### New blocking finding: the bounded-gap admission rejects a single-frame gap

`telegram-daemon.ts:9976` requires `msg.gap.fromSeq < msg.gap.toSeq`. The producer's range is **inclusive**: `SessionEventStream.replay` (`packages/coding-agent/src/sdk/host/events.ts:76-87`) emits `fromSeq: lastSeq + 1, toSeq: oldest - 1`. When exactly one frame has been evicted from the 256-frame ring — the ordinary steady-state overflow, since each append past capacity evicts exactly one — the legal tuple is `fromSeq === toSeq`. `sdk-host-wiring.test.ts:3310-3320` already pins that producer shape and asserts `fromSeq <= toSeq`.

Reproduced at head by taking your own passing test "a sequence-gap replay with a consistent identity resumes the exact endpoint authority" and changing nothing but the gap width:

```
gap: { kind: "sequence_gap", fromSeq: 2, toSeq: 5, ... }  → session resumes (your test)
gap: { kind: "sequence_gap", fromSeq: 2, toSeq: 2, ... }  → expect(sessions.get("S")).toBe(session)
                                                            received: undefined  → dropSession("invalid_replay")
```

Identity proof present, generation and cursor valid, socket authenticated — and the transport is still dropped. This is not a regression (the pre-fix predicate rejected every gap), but it means the recovery the PR is adding misses the most likely eviction shape and leaves those sessions in the same reconnect loop. Fix: `fromSeq <= toSeq` with the existing safe-integer and positive-bound checks, plus a daemon-level test at the 1..1 boundary. The over-permissive direction is already well fenced — forcing `gapAllowsRecovery = true` is caught by two tests — so only the boundary needs moving.

### Correcting my finding 1 (non-forum settlement) — it is broader than I wrote, not narrower

I originally justified it by pointing at `telegram-daemon.ts:435`. The authoritative justification is upstream: both strings come from **one shared gate**, TDLib `ForumTopicManager::is_forum` ([ForumTopicManager.cpp:1226-1252](https://github.com/tdlib/td/blob/master/td/telegram/ForumTopicManager.cpp#L1226-L1252)), which `toggle_forum_topic_is_closed` calls at line 961 with `allow_bots = false`:

- `"The chat is not a supergroup forum"` — line 1235, only the private-chat-with-a-forum-bot case (what your reproduction happened to hit).
- `"The chat is not a forum"` — line 1252, the fallthrough for **every other** non-forum chat: a basic group, a non-forum supergroup, a non-forum channel.

So the phrasing you allowlisted is the *narrow* case and the one you don't handle is the *common* one — a plain supergroup that was never a forum will return the short form from `closeForumTopic` and retry the archive forever. This is method-agnostic: it is the same gate for create and close. Finding stands, and the fix is the same: accept `the chat is not a (?:supergroup )?forum`.

### Factual correction in my own review

"`#recoverTopicBinding` is called from exactly one place" is wrong — it has two callers, `telegram-daemon.ts:7248` (`#updateLogicalSessionForThreadedFrame`) and `:10014` (the replay barrier). The conclusion is unaffected: the 7248 path does not redial a retained same-ID transport, so it is not an alternative convergence route, and deleting `recoveryRejectedAfterArchive` still leaves all 616 focused tests green while my convergence probe fails. Read "permanently mute" as "no deterministic same-endpoint convergence path absent a socket close, endpoint change, or unrelated reconnect event".

### Refinements to the non-blocking items

- Finding 4 (replay queue): the right fix is to **discard** `session.replayQueue` on the keep-attached branch, not drain it — the rejected lease makes the post-barrier gate at `:10097` reject routing anyway, so draining would be dead work. Withdraw my ownership/lease recheck suggestion: the branch performs no effect after its synchronous state update.
- New P3: `parseTopicRegistryState` validates `retiredTopics` recursively but throws away the normalized result (`topic-registry.ts:349-356`), and `TopicRegistry.load` copies the original records (`:443-447`), so a stale `disconnectGraceExpiresAt` in retired history is never repaired on round trip. Non-routable, so cosmetic today, but it makes the new repair asymmetric.

Updated blocker list: (1) inclusive one-event gap, (2) short-form non-forum settlement, (3) missing convergence regression test. Everything else is P3.

—
*[repo owner's gaebal-gajae (clawdbot) 🦞]*

**Comment by chiznoir** [comment]
Updated on current `dev@692393092` with the requested review fixes: inclusive single-event replay gaps (`fromSeq <= toSeq`) plus a 1..1 daemon regression, both Telegram non-forum archive phrasings, mutation-sensitive archive-settlement redial/successor delivery coverage, fenced replay-queue discard, and normalized retired-topic history without caller mutation. Verification at head `71964f704`: 667 focused Telegram/generation tests pass (2764 assertions), the two former dev regressions pass (39 tests), changed-file Biome and coding-agent TypeScript pass, generation guard validates, and a 180-second local source-daemon dogfood retained a fresh heartbeat with one live endpoint and delivered to Telegram. Full aggregate check/test were intentionally not rerun in this pass.

**Comment by chiznoir** [comment]
CI is now green at exact head `71964f704`: affected-path validation, coding-agent shard/check/build jobs, both Telegram suites, generation guard, Windows Telegram safety, native builds, root check, evidence producer, and virtual integration all pass. The requested blockers are addressed at this head, including the inclusive 1..1 replay-gap boundary, both non-forum settlement phrasings, mutation-sensitive archive convergence/redial coverage, fenced replay-queue discard, and normalized retired-topic history. Requesting exact-head re-review.

**Review by probepark — COMMENTED** [review]
## Stale-review refresh — exact head `71964f7046b1055889e449e731e170a35d771eca`

I re-read the prior review at `b5eea249cf726146dc8d72567e1c123c6d37c78d` and concentrated on the two follow-up commits. The previously reported non-forum alias, convergence-test, replay-queue, redundant-field, and retired-history issues are addressed. One new authority-boundary issue remains.

### Major

1. **A malformed sequence-gap claim can now authorize recovery without proving it describes the replay request.** `packages/coding-agent/src/sdk/bus/telegram-daemon.ts:9969-9980`

The new admission predicate checks only that `fromSeq <= toSeq`, `resyncQueries` is an array of strings, and some identity frame exists. It does not require the gap to start at the sequence this socket requested, require `toSeq <= lastSeq`, or ensure retained replay events sit above the conceded range. The canonical replay consumer performs those checks before conceding a gap (`chat-daemon-runtime.ts:811-833`) because otherwise a malformed/stale response can fence off sequences the host did not actually concede. Here, a response such as `lastSeq: 10`, `gap: { fromSeq: 100, toSeq: 200, ... }` plus a matching identity is accepted, marks the transport trusted, and later persists cursor 10.

The added test at `packages/coding-agent/test/notifications-telegram-daemon.test.ts:20662-20707` covers only one valid `1..1` gap. The rejection table at `:20718-20724` rejects a multi-sequence gap only because it has no identity, so it does not exercise contradictory bounds with otherwise valid identity proof. Validate the gap against the persisted/requested cursor and `lastSeq` (and retained event sequence bounds) before using it as admission proof, and add malformed-boundary cases.

### Minor

2. **The new regression adds repository-contract violations in test code.** `packages/coding-agent/test/notifications-telegram-daemon.test.ts:19008-19015,19044`

The test adds synchronous `fs.mkdirSync`/`fs.writeFileSync` calls instead of the required Bun/async filesystem APIs, and adds `FakeWs as any` despite the checkout contract's no-`any` rule. Use the existing typed test helper/constructor shape and async/Bun file setup rather than adding new exceptions.

### Verification

- `notifications-topic-registry.test.ts`: 59 pass / 0 fail.
- `notifications-telegram-daemon.test.ts`: could not start in this checkout because `@gajae-code/utils/shell-config` is not linked; 0 tests executed. I did not install or mutate the worktree.
- No placeholder tests, bare `not.toThrow()`, or `mock.module()` were added in the reviewed delta.

gajae.pr-review-verdict.v1: needs-human

**Comment by Yeachan-Heo** [comment]
Exact-head CI attribution at `8b2165fa4321150ae36813cb35cf60a9d01c2663` against `dev@05d52ba0b27ee20d3138781b10fb2d9d0e2fe91e`:

- Telegram generation authority drift is repaired; the latest generation guard is no longer the blocker.
- shard-1 ran 2,381 tests: **2,311 pass / 68 skip / 2 fail**. Both failures are current-base README contracts removed by merged #4219:
  - `external controller integration docs > keeps the guide discoverable from top-level and embedded docs`
  - `external controller integration docs > documents Aside as an opt-in search/context sidecar only`
- root-check independently fails the same #4219 documentation regression: root README no longer states the `red-claw` dark and `blue-crab` light defaults.
- evidence-producer and aggregate failures are downstream consequences of those required jobs.

The shared documentation repair is owned by #4225 and is already in an active current-dev lane. No #4117 Telegram source change is justified by this CI result. Once #4225 lands, I will refresh/rebase this exact PR head, rerun CI, perform a fresh signed review, and carry it to terminal merge/close disposition.

—
*[repo owner's gaebal-gajae (clawdbot) 🦞]*

**Comment by Yeachan-Heo** [comment]
## Exact-head CI fully green — MERGE_READY

**Head:** `09660907766f9ca435be5107e1905554b5ddc9f3`
**Base:** `dev` (current: `53306b55`)
**CI run:** [31494797325](https://github.com/Yeachan-Heo/gajae-code/actions/runs/31494797325) — **success**

### What was rebased
The original 9-commit stack by `chiznoir` was rebased onto current `dev` (`53306b55`) to clear exact-base ancestry failures. The rebase resolved generation-number conflicts: dev took gen 60 (#4200 heartbeat-sidecar publish resilience), so this PR's fenced transport became gen 61 and replay-gap authority validation became gen 62. The generation manifest was regenerated and the generation guard passes.

A 10th commit applies the minimal base-wide Biome cleanup (`session-manager.ts` optional-chain, `websocket-transport.ts` formatting) — same drift blocker as #4238, unrelated to the telegram daemon fix.

### Review blockers addressed
Both previously-requested blockers are in the code at head:

1. **Inclusive single-event replay gap** (Yeachan-Heo): `fromSeq <= toSeq` — the ordinary steady-state 1..1 overflow case is admitted.
2. **Replay-gap authority proof** (probepark): `fromSeq === session.replaySinceSeq + 1`, `toSeq <= lastSeq`, `!retainedAtOrBelowGap`, and identity frames required — a gap must describe the exact requested cursor and retain no event inside the conceded range.

### All checks green at exact head
- ✓ shard-1-of-8 (2,381 tests)
- ✓ root-check
- ✓ check:@gajae-code/coding-agent (Biome + tsc)
- ✓ notifications-telegram-daemon.test.ts (567 tests)
- ✓ notifications-topic-registry.test.ts (59 tests)
- ✓ Telegram daemon generation guard
- ✓ Windows Telegram daemon safety
- ✓ evidence producer + affected path validation
- ✓ Virtual integration validation

gaebal-gajae
